xonsh/completers/path.py

```diff
@@ -151,6 +151,59 @@ def _quote_paths(paths, start, end):
     return out
 
 
+def joinpath(path):
+    if path is None:
+        return ''
+    if len(path) == 0:
+        return ''
+    if path == ('',):
+        return get_sep()
+    elif path[0] == '':
+        return get_sep() + _normpath(os.path.join(*path))
+    else:
+        return _normpath(os.path.join(*path))
+
+
+def splitpath(path):
+    path = _normpath(path)
+    if path.startswith(get_sep()):
+        pre = ('', )
+    else:
+        pre = ()
+    return pre + _splitpath(path, ())
+
+
+def _splitpath(path, sofar=()):
+    folder, path = os.path.split(path)
+    if path == "":
+        return sofar[::-1]
+    elif folder == "":
+        return (sofar + (path, ))[::-1]
+    else:
+        return _splitpath(folder, sofar + (path, ))
+
+
+def expanded_match(ref, typed):
```

Comment:
I can certainly add some docs!  But what @gforsyth said is correct.  This implements _subsequence_ matching, so `pt` will match anything that has `p` and `t` in it (in that order), regardless of where they show up in the string.

I haven't actually played much with fish/zsh, but I think this is what they do (the `expanded_match` function I nabbed from @GuillaumeLeclerc (#903) and built from there).